File: base/grid_layout.cc
#include "base/grid_layout.h"

#include <stdexcept>

using namespace base;
using namespace std;

//----------------------------------------------------------------

grid_layout::grid_layout()
	: nr_fields_(0)
{
	new_row();
}

void
grid_layout::render(ostream &out) const
{
	vector<unsigned> widths;
	calc_field_widths(widths);

	grid::const_iterator row;
	bool newline_needed = false;

	for (row = grid_.begin(); row != grid_.end(); ++row) {
		row::const_iterator col;
		unsigned i;

		if (newline_needed) {
			out << "\n";
			newline_needed = false;
		}

		for (col = row->begin(), i = 0; col != row->end(); ++col, ++i) {
			out << justify(widths[i], *col) << " ";
			newline_needed = true;
		}
	}
}

void
grid_layout::new_row()
{
	grid_.push_back(row());
}

grid_layout::row const &
grid_layout::current_row() const
{
	return grid_.back();
}

grid_layout::row &
grid_layout::current_row()
{
	return grid_.back();
}

void
grid_layout::push_field(string const &s)
{
	current_row().push_back(s);
	nr_fields_ = max<unsigned>(nr_fields_, current_row().size());
}

void
grid_layout::calc_field_widths(vector<unsigned> &widths) const
{
	widths.resize(nr_fields_, 0);

	grid::const_iterator row;
	for (row = grid_.begin(); row != grid_.end(); ++row) {
		row::const_iterator col;
		unsigned i;
		for (col = row->begin(), i = 0; col != row->end(); ++col, ++i)
			widths[i] = max<unsigned>(widths[i], col->length());
	}
}

string
grid_layout::justify(unsigned width, string const &txt) const
{
	if (txt.length() > width)
		throw runtime_error("string field too long, internal error");

	string result(width - txt.length(), ' ');
	result += txt;
	return result;
}

//----------------------------------------------------------------

File: src/grid_layout.rs
use anyhow::Result;
use std::io::Write;
use std::string::String;
use std::vec::Vec;

pub struct GridLayout {
    grid: Vec<Vec<String>>,
    current: Vec<String>,
    max_columns: usize,
}

impl GridLayout {
    pub fn new() -> GridLayout {
        GridLayout {
            grid: Vec::new(),
            current: Vec::new(),
            max_columns: 0,
        }
    }

    pub fn new_with_size(rows: usize, columns: usize) -> GridLayout {
        GridLayout {
            grid: Vec::with_capacity(rows),
            current: Vec::with_capacity(columns),
            max_columns: columns,
        }
    }

    pub fn field(&mut self, s: String) {
        self.current.push(s)
    }

    pub fn new_row(&mut self) {
        if self.current.len() > self.max_columns {
            self.max_columns = self.current.len();
        }
        let mut last = Vec::with_capacity(self.max_columns);
        std::mem::swap(&mut self.current, &mut last);
        self.grid.push(last);
    }

    fn calc_field_widths(&self) -> Result<Vec<usize>> {
        let mut widths = vec![0; self.max_columns];
        for row in self.grid.iter() {
            for (col, width) in row.iter().zip(widths.iter_mut()) {
                *width = std::cmp::max(*width, col.len());
            }
        }
        Ok(widths)
    }

    // right align the string
    fn push_justified(buf: &mut String, s: &str, width: usize) {
        for _ in 0..width - s.len() {
            buf.push(' ');
        }
        buf.push_str(s);
        buf.push(' ');
    }

    pub fn render(&self, w: &mut dyn Write) -> Result<()> {
        let widths = self.calc_field_widths()?;

        for row in self.grid.iter() {
            let mut line = String::new();
            for (col, width) in row.iter().zip(widths.iter()) {
                Self::push_justified(&mut line, col.as_str(), *width);
            }
            line.push('\n');
            w.write_all(line.as_bytes())?;
        }

        Ok(())
    }
}

impl Default for GridLayout {
    fn default() -> Self {
        Self::new()
    }
}
